### ISSUE TITLE
about 'conf-thres' difference between val.py and detect.py

### ISSUE THREAD

**Issue by anonymoussss** [ultralytics/yolov5#4464]
## ❔Question
Hi, I have a question: why choose such a small parameter as small as 0.001 on 'conf-thres' in val.py? during detect, this parameter is set to 0.25.  
When saving a best performing model, does using a very low conf_thres benefit the final mAP performance?

## Additional context
train.py line347-365

        # DDP process 0 or single-GPU
        if rank in [-1, 0]:
            # mAP
            ema.update_attr(model, include=['yaml', 'nc', 'hyp', 'gr', 'names', 'stride', 'class_weights'])
            final_epoch = epoch + 1 == epochs
            if not opt.notest or final_epoch:  # Calculate mAP
                wandb_logger.current_epoch = epoch + 1
                results, maps, times = test.test(data_dict,
                                                 batch_size=batch_size * 2,
                                                 imgsz=imgsz_test,
                                                 model=ema.ema,
                                                 single_cls=opt.single_cls,
                                                 dataloader=testloader,
                                                 save_dir=save_dir,
                                                 verbose=nc < 50 and final_epoch,
                                                 plots=plots and final_epoch,
                                                 wandb_logger=wandb_logger,
                                                 compute_loss=compute_loss,
                                                 is_coco=is_coco)

**Comment by anonymoussss** [comment]
I am using my own custom dataset, and I try to change the val parameter at the end of an epoch in train.py from [conf_thres=0.001, iou_thres=0.6]  to [conf_thres=0.2, iou_thres=0.2], the test parameter from [conf_thres=0.001, iou_thres=0.7]  to [conf_thres=0.2, iou_thres=0.2].  The mAP performance did drop by 1 to 2 points in different model cfgs, could u tell me what might be the reason for this?

**Comment by glenn-jocher** [comment]
@anonymoussss 👋 Hello, thank you for asking about the differences between [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py), [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) and [val.py](https://github.com/ultralytics/yolov5/blob/master/val.py) in YOLOv5 🚀.

These 3 files are designed for different purposes and utilize different dataloaders with different settings. train.py dataloaders are designed for a speed-accuracy compromise, val.py is designed to obtain the best mAP on a validation dataset, and detect.py is designed for best real-world inference results. A few important aspects of each:

## `train.py`

- **trainloader** LoadImagesAndLabels(): designed to load train dataset images and labels. Augmentation capable and enabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/train.py#L188-L192
- **testloader** LoadImagesAndLabels(): designed to load val dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/train.py#L199-L202
- image size: `640`
- rectangular inference: `False`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `None`

## `val.py`

- **dataloader** LoadImagesAndLabels(): designed to load train, val, test dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/e88e8f7a988662fb2d613e1aca3ae89214c84084/val.py#L145-L146
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `0.5 * maximum stride`

## `detect.py`

- **dataloaders** (multiple): designed for loading multiple types of media (images, videos, globs, directories, streams). https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/detect.py#L46-L53
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L152): `0.25` 
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L153): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L415): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L45-L54): `None`

## YOLOv5 PyTorch Hub Inference

`models.autoShape()` class used for image loading, preprocessing, inference and NMS. For more info see [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/models/common.py#L225-L250

- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.25 `
- [iou threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L288): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `None`

**Comment by anonymoussss** [comment]
I see. thx :)
